argmin/src/solver/brent/brentopt.rs

```diff
@@ -0,0 +1,292 @@
+//! a function of one variable", _Algorithms for minimization without
+//! derivatives_, Richard P. Brent, 1973, Prentice-Hall.
+
+/// Implementation of Brent's optimization method,
+use crate::core::{
+    ArgminFloat, Error, IterState, Operator, Problem, Solver, State, TerminationReason, KV,
+};
+#[cfg(feature = "serde1")]
+use serde::{Deserialize, Serialize};
+
+/// Brent's method
+///
+/// A minimization algorithm combining parabolic interpolation and the
+/// golden-section method.  It has the reliability of the golden-section
+/// method, but can be faster thanks to the parabolic interpolation steps.
+///
+/// # References:
+///
+/// "An algorithm with guaranteed convergence for finding a minimum of
+/// a function of one variable", _Algorithms for minimization without
+/// derivatives_, Richard P. Brent, 1973, Prentice-Hall.
+#[derive(Clone)]
+#[cfg_attr(feature = "serde1", derive(Serialize, Deserialize))]
+pub struct BrentOpt<F> {
+    /// relative tolerance
+    eps: F,
+    /// absolute tolerance
+    t: F,
+    /// left boundary of current interval
+    a: F,
+    /// right boundary of current interval
+    b: F,
+    /// last point where f was evaluated
+    u: F,
+    /// previous value of w
+    v: F,
+    /// point with the current second lowest value of f
+    w: F,
+    /// point with the current lowest value of f
+    x: F,
+    /// value of f in v
+    fv: F,
+    /// value of f in w
+    fw: F,
+    /// value of f in x
+    fx: F,
+    /// value of p/q in the second last step
+    e: F,
+    /// value of p/q in the last step
+    d: F,
+    /// (3-sqrt(5)) / 2
+    c: F,
+}
+
+impl<F: ArgminFloat> BrentOpt<F> {
+    /// Constructor
+    ///
+    /// The values `min` and `max` must bracket the minimum of the function.
+    pub fn new(min: F, max: F) -> Self {
+        BrentOpt {
+            eps: F::epsilon().sqrt(),
+            t: F::from_f64(1e-5).unwrap(),
+            a: min,
+            b: max,
+            u: F::nan(),
+            v: F::nan(),
+            w: F::nan(),
+            x: F::nan(),
+            fv: F::nan(),
+            fw: F::nan(),
+            fx: F::nan(),
+            e: F::zero(),
+            d: F::zero(),
+            c: F::from_f64((3f64 - 5f64.sqrt()) / 2f64).unwrap(),
+        }
+    }
+
+    /// Set the tolerance to the value required.
+    ///
+    /// The algorithm will return an approximation `x` of a local
+    /// minimum of the function, with an accuracy smaller than `3 tol`,
+    /// where `tol = eps*abs(x) + t`.
+    /// It is useless to set `eps` to less than the square root of the
+    /// machine precision (`F::epsilon().sqrt()`), which is its default
+    /// value.  The default value of `t` is `1e-5`.
+    pub fn set_tolerance(mut self, eps: F, t: F) -> Self {
+        self.eps = eps;
+        self.t = t;
+        self
+    }
+}
+
+impl<O, F> Solver<O, IterState<F, (), (), (), F>> for BrentOpt<F>
+where
+    O: Operator<Param = F, Output = F>,
```

Comment:
I just noticed that you require `O` to implement `Operator` here (probably because `BrentRoot` does the same). I was wondering if `CostFunction` would make more sense in this context for both Brent methods (even though it essentially boils down to the same thing because both Brent methods optimize a single number). I got this idea because I noticed that you store the result of `apply` as cost function value in `state`.

Reply:
You are right, I took this from `BrentRoot` without really asking myself why. However I have to say that I am not sure of the difference between `Operator` and `CostFunction`, they look very similar to me… Would you mind elaborating on when to use one rather than the other? I think that I see a pattern in the code where `CostFunction`'s `Output` is intended to be scalar whereas `Operator`'s `Output` not necessarily, is that correct?

Reply:
I must have changed `ArgminOp` to `Operator` in `BrentRoot` without giving it much thought. `Operator` is meant for cases where one would for instance need to apply a matrix to a vector (such as in `ConjugateGradient`). The returned type would then be some sort of vector. `CostFunction` on the other hand always returns a scalar. I'm currently in the process of rewriting the documentation and I definitely need to make this more clear.

Reply:
I see, thanks! I replaced `Operator` with `CostFunction` in both cases.